File: python/be/api/venues/mini.py
"""HTTP client for mini-api (services/mini-api). Mirror of node/be/api/venues/mini.js.

A response is returned whole -- status, lower-cased headers, parsed body -- so
a step can assert on any of them. Only a transport failure is ApiUnreachable
(grades Blocked); a 4xx/5xx is an answer, and often the answer under test.

The service rate-limits per client address. One scenario measures that on
purpose; every other scenario must not trip it by accident, so the client
watches X-RateLimit-Remaining and, when the budget is nearly spent, waits for
the window to pass before the next request. A measurement control, not a
retry: no response is ever discarded.
"""

from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request

BASE = os.environ.get("MINI_API_URL", "http://127.0.0.1:8787").rstrip("/")
ADMIN_TOKEN = os.environ.get("MINI_API_ADMIN_TOKEN", "local-admin-token")
WINDOW_MS = int(os.environ.get("MINI_API_RATE_WINDOW_MS", "10000"))


class ApiUnreachable(Exception):
    pass


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, *args, **kwargs):  # noqa: D401 - never follow
        return None


_opener = urllib.request.build_opener(_NoRedirect)


class MiniApi:
    def __init__(self, base: str = BASE) -> None:
        self.base = base
        self.remaining: int | None = None

    def request(self, method: str, path: str, token: str | None = None, origin: str | None = None, body=None, headers: dict | None = None, throttle: bool = True) -> dict:
        if throttle and self.remaining is not None and self.remaining <= 3:
            time.sleep(WINDOW_MS / 1000 + 0.2)
            self.remaining = None
        h = dict(headers or {})
        if token:
            h["Authorization"] = "Bearer " + token
        if origin is not None:
            h["Origin"] = origin
        data = None
        if body is not None:
            h["Content-Type"] = "application/json"
            data = json.dumps(body).encode()
        req = urllib.request.Request(self.base + path, data=data, headers=h, method=method)
        try:
            with _opener.open(req, timeout=15) as res:
                status, hdrs, text = res.status, res.headers, res.read().decode("utf-8", "replace")
        except urllib.error.HTTPError as err:
            status, hdrs, text = err.code, err.headers, err.read().decode("utf-8", "replace")
        except (urllib.error.URLError, TimeoutError, OSError) as err:
            raise ApiUnreachable(f"mini-api at {self.base} did not answer {method} {path}: {err}") from err
        try:
            parsed = json.loads(text) if text else None
        except json.JSONDecodeError:
            parsed = None
        out = {"status": status, "headers": {k.lower(): v for k, v in hdrs.items()}, "body": parsed, "text": text}
        rem = out["headers"].get("x-ratelimit-remaining")
        if rem is not None:
            self.remaining = int(rem)
        return out

    def get(self, path: str, **kw) -> dict:
        return self.request("GET", path, **kw)

    def post(self, path: str, body, **kw) -> dict:
        return self.request("POST", path, body=body, **kw)

    def set_indexer_paused(self, paused: bool) -> dict:
        r = self.post("/admin/indexer", {"paused": paused}, token=ADMIN_TOKEN)
        if r["status"] != 200:
            raise RuntimeError(f"could not set indexer paused={paused}: HTTP {r['status']} {r['text']}")
        return r

    def rewind_indexer(self, block: int) -> dict:
        r = self.post("/admin/indexer", {"rewind": block}, token=ADMIN_TOKEN)
        if r["status"] != 200:
            raise RuntimeError(f"could not rewind indexer to {block}: HTTP {r['status']} {r['text']}")
        return r

    def mint_token(self, scope: str, subject: str | None = None, ttl: int | None = None) -> str:
        body = {"scope": scope}
        if subject:
            body["subject"] = subject
        if ttl:
            body["ttl"] = ttl
        r = self.post("/auth/token", body, token=ADMIN_TOKEN)
        if r["status"] != 201:
            raise RuntimeError(f"could not mint a {scope} token: HTTP {r['status']} {r['text']}")
        return r["body"]["token"]

File: node/be/api/venues/mini.js
'use strict';

/**
 * HTTP client for mini-api (services/mini-api), the REST layer over the
 * indexer's store. Node's built-in fetch; no library.
 *
 * A response is returned whole -- status, lower-cased headers, parsed body --
 * so a step can assert on any of them. Only a transport failure is
 * ApiUnreachable (grades Blocked); a 4xx/5xx is an answer, and often the
 * answer under test.
 *
 * The service rate-limits per client address (60 per 10 s by default). One
 * scenario measures that on purpose; every other scenario must not trip it by
 * accident, so the client watches X-RateLimit-Remaining and, when the budget
 * is nearly spent, waits for the window to pass before the next request.
 * That is a measurement control, not a retry: no response is ever discarded.
 */

const BASE = (process.env.MINI_API_URL || 'http://127.0.0.1:8787').replace(/\/+$/, '');
const ADMIN_TOKEN = process.env.MINI_API_ADMIN_TOKEN || 'local-admin-token';
const WINDOW_MS = Number(process.env.MINI_API_RATE_WINDOW_MS || 10_000);

class ApiUnreachable extends Error {}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

class MiniApi {
  constructor(base = BASE) {
    this.base = base;
    this.remaining = null; // last X-RateLimit-Remaining seen
  }

  /** @returns {Promise<{status:number, headers:Object, body:any, text:string}>} */
  async request(method, path, { token, origin, body, headers = {}, throttle = true } = {}) {
    if (throttle && this.remaining !== null && this.remaining <= 3) {
      await sleep(WINDOW_MS + 200);
      this.remaining = null;
    }
    const h = { ...headers };
    if (token) h.authorization = 'Bearer ' + token;
    if (origin !== undefined) h.origin = origin;
    let init = { method, headers: h };
    if (body !== undefined) { h['content-type'] = 'application/json'; init.body = JSON.stringify(body); }
    let res;
    try {
      res = await fetch(this.base + path, init);
    } catch (err) {
      throw new ApiUnreachable('mini-api at ' + this.base + ' did not answer ' + method + ' ' + path + ': ' + (err.cause && err.cause.message ? err.cause.message : err.message));
    }
    const text = await res.text();
    let parsed = null;
    try { parsed = text ? JSON.parse(text) : null; } catch { parsed = null; }
    const out = { status: res.status, headers: Object.fromEntries([...res.headers.entries()].map(([k, v]) => [k.toLowerCase(), v])), body: parsed, text };
    const rem = out.headers['x-ratelimit-remaining'];
    if (rem !== undefined) this.remaining = Number(rem);
    return out;
  }

  get(path, opts) { return this.request('GET', path, opts); }
  post(path, body, opts = {}) { return this.request('POST', path, { ...opts, body }); }

  /** Admin lever: freeze or resume the indexer. */
  async setIndexerPaused(paused) {
    const r = await this.post('/admin/indexer', { paused }, { token: ADMIN_TOKEN });
    if (r.status !== 200) throw new Error('could not set indexer paused=' + paused + ': HTTP ' + r.status + ' ' + r.text);
    return r;
  }

  /** Admin lever: forget everything after `block` so the next sync replays the range. */
  async rewindIndexer(block) {
    const r = await this.post('/admin/indexer', { rewind: block }, { token: ADMIN_TOKEN });
    if (r.status !== 200) throw new Error('could not rewind indexer to ' + block + ': HTTP ' + r.status + ' ' + r.text);
    return r;
  }

  /** Mint a scoped token with the admin token. */
  async mintToken(scope, subject = null, ttl = null) {
    const body = { scope };
    if (subject) body.subject = subject;
    if (ttl) body.ttl = ttl;
    const r = await this.post('/auth/token', body, { token: ADMIN_TOKEN });
    if (r.status !== 201) throw new Error('could not mint a ' + scope + ' token: HTTP ' + r.status + ' ' + r.text);
    return r.body.token;
  }
}

module.exports = { MiniApi, ApiUnreachable, BASE, ADMIN_TOKEN, WINDOW_MS, sleep };
